addons/api/addon/handlers/indexed-db-handler.js

```diff
@@ -0,0 +1,119 @@
+import { modelIndexes } from 'api/services/indexed-db';
+import { inject as service } from '@ember/service';
+import { getOwner, setOwner } from '@ember/application';
+import { pluralize } from 'ember-inflector';
+
+/**
+ * Handler to sit in front of the API request layer
+ * so we can handle any caching with indexedDb first
+ */
+export default class IndexedDbHandler {
+  @service indexedDb;
+
+  constructor(context) {
+    setOwner(this, getOwner(context));
+  }
+
+  async request(context, next) {
+    switch (context.request.op) {
+      case 'query': {
+        const { store, data } = context.request;
+        const { type, query, options: { pushToStore = true } = {} } = data;
+        const supportedModels = Object.keys(modelIndexes);
+        const { db: indexedDb } = this.indexedDb ?? {};
+        const pluralizedType = pluralize(type);
+
+        // Go through normal flow if we don't yet support the model
+        // or if we don't have an indexedDb instance
+        if (!supportedModels.includes(pluralizedType) || !indexedDb) {
+          return next(context.request);
+        }
+
+        const listToken = await indexedDb.tokens.get(pluralizedType);
+        const adapter = store.adapterFor(type);
+        const schema = store.modelFor(type);
+
+        let payload;
+        try {
+          payload = await adapter.query(store, schema, {
+            ...query,
+            scope_id: 'global',
+            recursive: true,
+            list_token: listToken?.token,
+          });
+        } catch (e) {
+          // If we get any error that's not a rate limiting error,
+          // assume the token is not good anymore. We'll delete the token and
+          // clear the DB and try again without a token.
+          // TODO: Probably should check for 400 and bad token error specifically instead?
+          if (e?.errors[0].status !== 429) {
+            await indexedDb[pluralizedType].clear();
+            await indexedDb.tokens.delete(pluralizedType);
+
+            payload = await adapter.query(store, schema, {
+              ...query,
+              scope_id: 'global',
+              recursive: true,
+              list_token: null,
+            });
+          }
+        }
+
+        // Store the token we just got back from the payload
+        await indexedDb.tokens.put({
+          id: pluralizedType,
+          token: payload.list_token,
+        });
```

Comment:
perhaps we should check if there even is a token first in the case of an empty response?